automation/src/test/java/org/wso2/carbon/testgrid/automation/executor/JMeterExecutorTest.java

```diff
@@ -0,0 +1,346 @@
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.awaitility.Awaitility;
+import org.testng.Assert;
+import org.testng.annotations.AfterTest;
+import org.testng.annotations.BeforeTest;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Parameters;
+import org.testng.annotations.Test;
+import org.wso2.carbon.testgrid.automation.TestAutomationException;
+import org.wso2.carbon.testgrid.automation.executor.util.ZipUtil;
+import org.wso2.carbon.testgrid.common.Deployment;
+import org.wso2.carbon.testgrid.common.Host;
+import org.wso2.carbon.testgrid.common.Port;
+import org.wso2.carbon.testgrid.common.constants.TestGridConstants;
+import org.wso2.carbon.testgrid.common.util.EnvironmentUtil;
+
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.FileReader;
+import java.io.IOException;
+import java.net.ConnectException;
+import java.net.Socket;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.attribute.PosixFilePermission;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.Callable;
+import java.util.concurrent.TimeUnit;
+
+import static org.hamcrest.CoreMatchers.equalTo;
+
+/**
+ * Test class to test the functionality of the {@link JMeterExecutor} class.
+ *
+ * @since 1.0.0
+ */
+public class JMeterExecutorTest {
+
+    private static final Log log = LogFactory.getLog(JMeterExecutorTest.class);
+    private final String productName = "wso2is-5.3.0";
+    private String distributionLocation;
+
+    @BeforeTest
+    public void setUp() throws IOException {
+        // Unzip distribution
+        ClassLoader classLoader = getClass().getClassLoader();
+        URL resource = classLoader.getResource("distributions");
+        Assert.assertNotNull(resource);
+        distributionLocation = Paths.get(resource.getPath(), productName).toAbsolutePath().toString();
+
+        Path distributionZipPath = Paths.get(resource.getPath(), productName + ".zip").toAbsolutePath();
+        ZipUtil.unzip(distributionZipPath.toString(), resource.getPath());
+
+        // Set file permissions
+        if (SystemUtils.IS_OS_LINUX || SystemUtils.IS_OS_MAC) {
+            Path shellScriptPath = Paths.get(distributionLocation, "bin", "wso2server.sh")
+                    .toAbsolutePath();
+            setPermission(shellScriptPath);
+        }
+
+        // Start server
+        String execFile = SystemUtils.IS_OS_WINDOWS ?
+                          Paths.get(distributionLocation, "bin", "wso2server.bat").toAbsolutePath().toString() :
+                          Paths.get(distributionLocation, "bin", "wso2server.sh").toAbsolutePath().toString();
+        ProcessBuilder processBuilder = new ProcessBuilder(execFile);
+        processBuilder.start();
```

Comment:
Unit tests should only do functional testing. If you want to test how the composed server functions, then we should write integration tests instead.

For your usecase, it seems you need a WSO2 server to verify jmeter. I would suggest that we mock the wso2server https endpoint. We can use Mockito.

Reply:
Mocked via an external server (https://www.mocky.io/).